File: frontend/iosApp/UILogics/Sources/UILogics/Screens/LoginViewModel.swift
import Foundation
import Domains
import UseCases

@MainActor
public final class LoginViewModel: ObservableObject {
    private let loginUseCase: LoginUseCaseProtocol
    private let onSuccess: (AuthSession) -> Void

    @Published public var username: String = ""
    @Published public var password: String = ""
    @Published public var loginState: LoginState = .idle

    public var continueAsItem: ContinueAsUIModel?

    public init(
        loginUseCase: LoginUseCaseProtocol,
        onSuccess: @escaping (AuthSession) -> Void,
        continueAsItem: ContinueAsUIModel? = nil
    ) {
        self.loginUseCase = loginUseCase
        self.onSuccess = onSuccess
        self.continueAsItem = continueAsItem
    }

    public func login() async {
        loginState = .loading

        let result = await loginUseCase.login(
            username: username,
            password: password
        )

        switch result {
        case .success(let session):
            loginState = .idle
            onSuccess(session)
        case .failure(let error):
            loginState = .error(message: errorMessage(for: error))
        }
    }

    private func errorMessage(for error: LoginUseCaseModel.LoginResult.Error) -> String {
        switch error {
        case .invalidCredentials:
            return "Invalid username or password"
        case .networkError:
            return "Network error"
        case .serverError:
            return "Server error"
        case .unknown:
            return "Unknown error"
        }
    }
}

extension LoginViewModel {
    public struct ContinueAsUIModel {
        public let userName: String
        public let avatarUrl: URL?
        public let onTap: () -> Void

        public init(userName: String, avatarUrl: URL?, onTap: @escaping () -> Void) {
            self.userName = userName
            self.avatarUrl = avatarUrl
            self.onTap = onTap
        }
    }

    public enum LoginState {
        case idle
        case loading
        case error(message: String)

        public var isLoading: Bool {
            if case .loading = self { return true }
            return false
        }
    }
}

File: frontend/composeApp/src/androidMain/kotlin/com/example/memoriesapp/ui/uilogics/viewmodels/LoginViewModel.kt
package com.example.memoriesapp.ui.uilogics.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.memoriesapp.domain.AuthSession
import com.example.memoriesapp.domain.User
import com.example.memoriesapp.usecase.LoginError
import com.example.memoriesapp.usecase.LoginResult
import com.example.memoriesapp.usecase.LoginUseCase
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for login screen.
 */
class LoginViewModel(
    private val loginUseCase: LoginUseCase,
    val continueAsItem: ContinueAsItem? = null
) : ViewModel() {

    var username by mutableStateOf("")
        private set
    var password by mutableStateOf("")
        private set
    var loginState by mutableStateOf<LoginState>(LoginState.Idle)
        private set

    private val _navigationEvent = MutableSharedFlow<NavigationEvent>()
    val navigationEvent: SharedFlow<NavigationEvent> = _navigationEvent.asSharedFlow()

    fun onUsernameChange(value: String) {
        username = value
    }

    fun onPasswordChange(value: String) {
        password = value
    }

    fun login() {
        viewModelScope.launch {
            loginState = LoginState.Loading
            when (val result = loginUseCase.login(username, password)) {
                is LoginResult.Success -> {
                    loginState = LoginState.Idle
                    _navigationEvent.emit(NavigationEvent.LoginSuccess(result.session))
                }
                is LoginResult.Failure -> {
                    loginState = LoginState.Error(errorMessage(result.error))
                }
            }
        }
    }

    fun continueAsUser() {
        viewModelScope.launch {
            _navigationEvent.emit(NavigationEvent.ContinueAsUser)
        }
    }

    private fun errorMessage(error: LoginError): String {
        return when (error) {
            LoginError.INVALID_CREDENTIALS -> "Invalid username or password"
            LoginError.NETWORK_ERROR -> "Network error"
            LoginError.SERVER_ERROR -> "Server error"
            LoginError.UNKNOWN -> "Unknown error"
        }
    }

    data class ContinueAsItem(
        val userName: String,
        val avatarUrl: String?
    )

    sealed class LoginState {
        data object Idle : LoginState()
        data object Loading : LoginState()
        data class Error(val message: String) : LoginState()

        val isLoading: Boolean
            get() = this is Loading
    }

    sealed class NavigationEvent {
        data class LoginSuccess(val session: AuthSession) : NavigationEvent()
        data object ContinueAsUser : NavigationEvent()
    }
}
